diff --git a/mysys/mysys_priv.h b/mysys/mysys_priv.h
index adf2d39046a..e4bef7296e3 100644
--- a/mysys/mysys_priv.h
+++ b/mysys/mysys_priv.h
@@ -144,7 +144,10 @@ void my_error_unregister_all(void);
 #endif
 #endif
 
-#ifdef O_PATH
+// EDG: openat not supported, always use NOAT variant
+#if 1
+#define NOSYMLINK_FUNCTION_BODY(AT,NOAT) return NOAT;
+#elif 0
 #define HAVE_OPEN_PARENT_DIR_NOSYMLINKS
 const char *my_open_parent_dir_nosymlinks(const char *pathname, int *pdfd);
 #define NOSYMLINK_FUNCTION_BODY(AT,NOAT)                                \
diff --git a/sql/mysqld.cc b/sql/mysqld.cc
index 48a09d15755..b5e8b6f3e54 100644
--- a/sql/mysqld.cc
+++ b/sql/mysqld.cc
@@ -5510,7 +5510,7 @@ int mysqld_main(int argc, char **argv)
     init signals & alarm
     After this we can't quit by a simple unireg_abort
   */
-  start_signal_handler();				// Creates pidfile
+  (void)start_signal_handler; // EDG: signals not supported
 
   if (mysql_rm_tmp_tables() || acl_init(opt_noacl) ||
       my_tz_init((THD *)0, default_tz_name, opt_bootstrap))
diff --git a/storage/rocksdb/CMakeLists.txt b/storage/rocksdb/CMakeLists.txt
index c2fc8de8149..cff5ac101fa 100644
--- a/storage/rocksdb/CMakeLists.txt
+++ b/storage/rocksdb/CMakeLists.txt
@@ -121,12 +121,12 @@ else()
   SET(ATOMIC_EXTRA_LIBS)
 endif()
 
-MYSQL_ADD_PLUGIN(rocksdb ${ROCKSDB_SE_SOURCES} MODULE_ONLY STORAGE_ENGINE
+MYSQL_ADD_PLUGIN(rocksdb_se ${ROCKSDB_SE_SOURCES} MANDATORY STORAGE_ENGINE
                  MODULE_OUTPUT_NAME ha_rocksdb
                  LINK_LIBRARIES ${ATOMIC_EXTRA_LIBS}
                  COMPONENT rocksdb-engine)
 
-IF(NOT TARGET rocksdb)
+IF(NOT TARGET rocksdb_se)
   # Bail out if compilation with rocksdb engine is not requested
   RETURN()
 ENDIF()
@@ -197,7 +197,7 @@ TARGET_LINK_LIBRARIES(rocksdb_aux_lib ${ATOMIC_EXTRA_LIBS})
 #ENDIF()
 
 
-TARGET_LINK_LIBRARIES(rocksdb rocksdb_aux_lib)
+TARGET_LINK_LIBRARIES(rocksdb_se rocksdb_aux_lib)
 
 CHECK_FUNCTION_EXISTS(sched_getcpu  HAVE_SCHED_GETCPU)
 IF(HAVE_SCHED_GETCPU)
