diff --git a/storage/rocksdb/build_rocksdb.cmake b/storage/rocksdb/build_rocksdb.cmake
index 3a2297fd0f2..45285f09e63 100644
--- a/storage/rocksdb/build_rocksdb.cmake
+++ b/storage/rocksdb/build_rocksdb.cmake
@@ -3,7 +3,7 @@ if(POLICY CMP0042)
   cmake_policy(SET CMP0042 NEW)
 endif()
 
-SET(ROCKSDB_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/rocksdb)
+SET(ROCKSDB_SOURCE_DIR ${EDBDIR}/3rdparty/edgeless-rocksdb)
 
 INCLUDE_DIRECTORIES(
   ${CMAKE_CURRENT_BINARY_DIR}
@@ -295,7 +295,6 @@ set(ROCKSDB_SOURCES
         options/options_parser.cc
         options/options_sanity_check.cc
         port/stack_trace.cc
-        table/adaptive/adaptive_table_factory.cc
         table/block_based/block.cc
         table/block_based/block_based_filter_block.cc
         table/block_based/block_based_table_builder.cc
@@ -314,21 +313,12 @@ set(ROCKSDB_SOURCES
         table/block_based/partitioned_filter_block.cc
         table/block_based/uncompression_dict_reader.cc
         table/block_fetcher.cc
-        table/cuckoo/cuckoo_table_builder.cc
-        table/cuckoo/cuckoo_table_factory.cc
-        table/cuckoo/cuckoo_table_reader.cc
         table/format.cc
         table/get_context.cc
         table/iterator.cc
         table/merging_iterator.cc
         table/meta_blocks.cc
         table/persistent_cache_helper.cc
-        table/plain/plain_table_bloom.cc
-        table/plain/plain_table_builder.cc
-        table/plain/plain_table_factory.cc
-        table/plain/plain_table_index.cc
-        table/plain/plain_table_key_coding.cc
-        table/plain/plain_table_reader.cc
         table/sst_file_reader.cc
         table/sst_file_writer.cc
         table/table_properties.cc
@@ -415,6 +405,7 @@ set(ROCKSDB_SOURCES
         utilities/ttl/db_ttl_impl.cc
         utilities/write_batch_with_index/write_batch_with_index.cc
         utilities/write_batch_with_index/write_batch_with_index_internal.cc
+        file/encrypted_file.cc
 )
 
 
@@ -501,3 +492,12 @@ target_link_libraries(rocksdblib ${THIRDPARTY_LIBS} ${SYSTEM_LIBS})
 IF(CMAKE_CXX_COMPILER_ID MATCHES "GNU" OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   set_target_properties(rocksdblib PROPERTIES COMPILE_FLAGS "-fPIC -fno-builtin-memcmp -Wno-error")
 endif()
+
+# EDG: additional target config
+target_compile_definitions(rocksdblib PRIVATE NO_ALTERNATIVE_TABLES)
+set_target_properties(rocksdblib PROPERTIES CXX_STANDARD 17)
+
+# EDG: link libedgeless
+set(LIBEDG_NOTEST ON)
+add_subdirectory(${ROCKSDB_SOURCE_DIR}/libedgeless libedgeless)
+target_link_libraries(rocksdblib edgeless_o)
